core/services/gateway/gateway.go

```diff
@@ -0,0 +1,80 @@
+package gateway
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/job"
+)
+
+type Gateway interface {
+	job.ServiceCtx
+}
+
+type gateway struct {
+	codec    Codec
+	handlers map[string]Handler
+	connMgr  ConnectionManager
+	lggr     logger.SugaredLogger
+}
+
+func NewGatewayFromConfig(config *GatewayConfig, lggr logger.SugaredLogger) (Gateway, error) {
+	codec := &JsonRPCCodec{}
+
+	handlers := make(map[string]Handler)
+	donConnMgrs := make(map[string]DONConnectionManager)
+	for _, donConfig := range config.Dons {
+		donConfig := donConfig
+		if donConfig.DonId == "" {
+			return nil, errors.New("empty DON ID")
+		}
+		handler, err := NewHandler(donConfig.HandlerName)
+		if err != nil {
+			return nil, err
+		}
+		_, ok := handlers[donConfig.DonId]
+		if ok {
+			return nil, fmt.Errorf("duplicate DON ID %s", donConfig.DonId)
+		}
+		handlers[donConfig.DonId] = handler
+		donConnMgr := NewDONConnectionManager(&donConfig, handler, codec)
+		donConnMgrs[donConfig.DonId] = donConnMgr
+		err = handler.Init(donConnMgr, &donConfig)
+		if err != nil {
+			return nil, fmt.Errorf("error initializing handler: %w", err)
+		}
+	}
+	connMgr := NewConnectionManager(donConnMgrs, lggr)
+	return NewGateway(codec, handlers, connMgr, lggr), nil
+}
+
+func NewGateway(codec Codec, handlers map[string]Handler, connMgr ConnectionManager, lggr logger.SugaredLogger) Gateway {
+	return &gateway{
+		codec:    codec,
+		handlers: handlers,
+		connMgr:  connMgr,
+		lggr:     lggr,
+	}
+}
+
+func (g *gateway) Start(ctx context.Context) error {
+	g.lggr.Info("starting gateway")
```

Comment:
For Start/Stop, you need to use `utils.StartStopOnce`, see the usages in Core.

Reply:
Ah great, thanks! Also switched to multierr in Close().